lib/utils/batcher/batcher.go

```diff
@@ -0,0 +1,186 @@
+// processed together in a single operation. For example, when reacting to role
+// change events and check what users have been affected by this change.
+// Instead of fetching all the users for each role change event
+// you can aggregate roles change events and fetch and process all users only once
+// instead of fetching all the users from each role change event.
+//
+// Example:
+//
+//	   opts := []batcher.Option{
+//	       batcher.WithClock(u.clock),
+//	       batcher.WithWindow(time.Second * 10),
+//	       batcher.WithThreshold(100),
+//	   }
+//
+//		err := batcher.Run(ctx, eventChan, func(batch []Event) error {
+//			log.Printf("Processing %d events", len(batch))
+//			return processEvents(batch)
+//		}, opts...))
+//
+// The function returns when the context is canceled, returning ctx.Err(),
+// when fn returns an error, returning that error, or when the events channel is closed.
+func Run[T any](ctx context.Context, events <-chan T, fn func(batch []T) error, opts ...Option) error {
+	collector := New[T](opts...)
+	return trace.Wrap(collector.Run(ctx, events, fn))
+}
+
+// Option configures a Collector using the functional options pattern.
+type Option func(*config)
+
+// WithWindow sets the collection window duration.
+func WithWindow(d time.Duration) Option {
+	return func(c *config) {
+		if d > 0 {
+			c.window = d
+		}
+	}
+}
+
+// WithThreshold sets the maximum number of events per batch.
+// Default is unlimited (bounded only by window duration).
+func WithThreshold(n int) Option {
+	return func(c *config) {
+		if n > 0 {
+			c.threshold = n
+		}
+	}
+}
+
+// WithClock sets a custom clock for testing.
+// Default uses the system clock.
+func WithClock(clock clockwork.Clock) Option {
+	return func(c *config) {
+		if clock != nil {
+			c.clock = clock
+		}
+	}
+}
+
+type config struct {
+	window    time.Duration
+	threshold int
+	clock     clockwork.Clock
+}
+
+// New creates a new Collector with the given options.
+func New[T any](opts ...Option) *Collector[T] {
+	cfg := config{
+		window:    10 * time.Second,
+		threshold: 50,
+		clock:     clockwork.NewRealClock(),
+	}
+	for _, opt := range opts {
+		opt(&cfg)
+	}
+	return &Collector[T]{config: cfg}
+}
+
+// Collector batches events into time windows.
+type Collector[T any] struct {
+	config
+}
+
+// Run starts collecting events and calls fn for each batch.
+// It blocks until ctx is canceled or an error occurs.
+func (c *Collector[T]) Run(ctx context.Context, events <-chan T, fn func(batch []T) error) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+		}
+
+		batch, err := c.CollectBatch(ctx, events)
+		if err != nil {
+			return err
+		}
+
+		if len(batch) > 0 {
+			if err := fn(batch); err != nil {
+				return trace.Wrap(err)
+			}
+		}
+	}
+}
+
+// ErrChannelClosed is returned when the events channel is closed.
+var ErrChannelClosed = trace.BadParameter("events channel closed")
```

Comment:
I'm not convinced `BadParameter` is a good error class for this but either way, the `trace` construction functions will store the stack trace at call time, which means that the stack trace is from the implicit `init()` (which doesn't help anything) and won't get overwritten by `trace.Wrap`. In package-level error values (which are sort of a subpar pattern anyway), prefer constructing the `trace` errors manually so a `trace.Wrap` at the return site will do the correct thing.

```suggestion
var ErrChannelClosed = &trace.BadParameterError{Message: "events channel closed"}
```